Use block-level sampling for ANALYZE on heap tables.

hlinnaka opened the pull request:
The user-visible feature here is to speed up ANALYZE on heap tables, by
using the block-level sampling method from PostgreSQL, instead of the
scanning the whole table. This refactors the ANALYZE code, to make it
possible to call the upstream function to do the block-level sampling
again.
    
A secondary feature is that this enables the ANALYZE callback for Foreign
Data Wrappers, introduced in PostgreSQL 9.2. It was disabled in the 9.2
merge, because the function signature had been changed in GPDB, making
it cumbersome to call it. This patch reverts acquire-function's signature
to the same as in upstream, so the FDW callback should now work again.
    
This also moves the logic to decide which rels to analyze, based on
optimizer_analyze_root_partition or optimizer_analyze_midlevel_partition
GUCs and the ROOTPARTITION option, wholly to get_rel_oids() in vacuum.c.
It was mostly done there already, but there were some warts in analyze.c.
None of the code in analyze.c should now treat partitions differently
from normal inherited tables, which helps to minimize the diff vs
upstream. (Except that the HLL code still heavily uses partition-related
functions. TODO!)

hlinnaka commented:
This is still a bit rough in the edges, as explain in the fixme/todo comments. But I wanted to get feedback on the general direction this is going, and to let others know that I've been hacking on this.

bhuvnesh2703 commented on src/backend/commands/analyze.c:
```diff
@@ -638,6 +672,8 @@ do_analyze_rel(Relation onerel, VacuumStmt *vacstmt,
 	 * setting to the original user is required due to GPDB specific way of collecting samples
 	 * using query, but not required postgres since we do block sampling.
+	 *
+	 * FIXME: I don't understand this. Why do we differ from upstream, again?
```
In upstream, there is block sampling, so we don't query the table and run into the permission issues
as demonstrated in the below test case seen on gpdb due to which we had to put in this specific fix.
With your (this) commit, i think we don't we need it different than upstream except due to HLL fullscan.
Refer to the below scenario on regular tables, it used to fail earlier on GPDB but not in upstream due to block sampling vs querying the table. 
```
CREATE ROLE test_role;
CREATE SCHEMA test_schema;
CREATE TABLE test_schema.test_table(a INT, b VARCHAR);
INSERT INTO test_schema.test_table SELECT *, 'test' FROM generate_series(1, 1000);
ALTER TABLE test_schema.test_table OWNER TO test_role;
ANALYZE test_schema.test_table;
ERROR: permission denied for schema test_schema
LINE 1: ... when Ta.b is NULL then false else true end) from test_schem...
```

WIth your commit the above will work.
```
```

However, in case of HLL FULLSCAN on partitioned tables, we still query the table, and will run into the permission issue if we don't have the fix.
```
CREATE ROLE test_role;
CREATE SCHEMA test_schema;
CREATE TABLE test_schema.test_table(a INT, b int) partition by range(b) (start(0) end(5) every(1));
INSERT INTO test_schema.test_table SELECT i, i%5 FROM generate_series(1, 1000)i;
ALTER TABLE test_schema.test_table OWNER TO test_role;
pivotal=# ANALYZE FULLSCAN test_schema.test_table;
ERROR:  permission denied for schema test_schema
LINE 1: ...og_accum(a), pg_catalog.hyperloglog_accum(b) from test_schem...
```

hlinnaka replied:
Oh, I didn't notice the `SetUserIdAndSecContext` call after the call to *acquirefunc. I guess it's correct, then, but it's very subtle.. Commit 9c1b1ae3af made it much more subtle, because it moved the SetUserIdAndSecContext() to within the if() statement. I think that was a bad idea.

bhuvnesh2703 commented:
I see fixing the tests is a TODO, but putting a question for it.

```
DROP TABLE IF EXISTS foo;
CREATE TABLE foo (a int, b int, c text) PARTITION BY RANGE (b) (START (0) END (6) EVERY (3));
INSERT INTO foo select i, i%6, repeat('aaaa', 100000) FROM generate_series(1, 100)i;
ANALYZE foo;
SELECT * FROM pg_stats WHERE tablename like 'foo%' and attname = 'c' ORDER BY attname,tablename;
 schemaname |  tablename  | attname | inherited | null_frac | avg_width | n_distinct | most_common_vals | most_common_freqs | histogram_bounds | correlation | most_common_elems | most_common_elem_freqs | elem_count_histogram 
------------+-------------+---------+-----------+-----------+-----------+------------+------------------+-------------------+------------------+-------------+-------------------+------------------------+----------------------
 public     | foo         | c       | t         |         1 |         0 |          0 |                  |                   |                  |             |                   |                        | 
 public     | foo_1_prt_1 | c       | f         |         1 |         0 |          0 |                  |                   |                  |             |                   |                        | 
 public     | foo_1_prt_2 | c       | f         |         1 |         0 |          0 |                  |                   |                  |             |                   |                        | 
```

The column c is inserted with a large value (potentially greater than 1kb). I see that the avg_width is counted as 0, because we don't transfer these large datums as is, but replace them with null. Is that correct?
Shouldn't the null_frac be 0 as they are not real null?

bhuvnesh2703 commented:
Again from incremental_analyze test
```
 -- Test ANALYZE full scan HLL
@@ -1923,9 +1932,9 @@
 SELECT relname, relpages, reltuples FROM pg_class WHERE relname LIKE 'incr_analyze_test%' ORDER BY relname;
           relname          | relpages | reltuples
 ---------------------------+----------+-----------
- incr_analyze_test         |        4 |      1001
+ incr_analyze_test         |        1 |      1001
  incr_analyze_test_1_prt_1 |        1 |         1
- incr_analyze_test_1_prt_2 |        3 |      1000
+ incr_analyze_test_1_prt_2 |        2 |      1000
  incr_analyze_test_1_prt_3 |        1 |         0
  incr_analyze_test_1_prt_4 |        1 |         0
  incr_analyze_test_1_prt_5 |        1 |         0
```
should the relpages change and why?

hlinnaka commented:
Ok, I spent some time cleaning this up. A new version is included. This is now ready for review and pushing, so I've updated the title and description, and removed the "WIP" label.

Please take a new look.

hlinnaka commented:
@bhuvnesh2703, I fixed the regression test failures in `analyze` and `incr_analyze`. Some of the discrepancies were straightforward bugs, some were minor differences in how stats are collected and summed up. The bugs have been fixed, and the expected output has been changed for the differences that seem OK. The individual commit messages hopefully clarify those.

hlinnaka commented:
Reviewers: if this seems too much to review in one go, consider reviewing one of the smaller commits first. The could be pushed separately, if we can get +1 on those, before the big patch.

Eulerizeit commented:
@hlinnaka does making this work for AO/AOCO tables make sense?

hlinnaka commented:
> @hlinnaka does making this work for AO/AOCO tables make sense?

Sure, that'd be cool. But it's a separate task, after this.

I think we could use the "block directory" for AO tables for it. The block directory is currently only created if the table has indexes, but we could start creating it unconditionally. Shouldn't add too much overhead. But the patch will need to deal with all that, so it's not going to be a tiny patch.

hlinnaka commented:
> Again from incremental_analyze test
> 
> ```
>  -- Test ANALYZE full scan HLL
> @@ -1923,9 +1932,9 @@
>  SELECT relname, relpages, reltuples FROM pg_class WHERE relname LIKE 'incr_analyze_test%' ORDER BY relname;
>            relname          | relpages | reltuples
>  ---------------------------+----------+-----------
> - incr_analyze_test         |        4 |      1001
> + incr_analyze_test         |        1 |      1001
>   incr_analyze_test_1_prt_1 |        1 |         1
> - incr_analyze_test_1_prt_2 |        3 |      1000
> + incr_analyze_test_1_prt_2 |        2 |      1000
>   incr_analyze_test_1_prt_3 |        1 |         0
>   incr_analyze_test_1_prt_4 |        1 |         0
>   incr_analyze_test_1_prt_5 |        1 |         0
> ```
> 
> should the relpages change and why?

Relpages is now calculated slightly differently. Before, ANALYZE ran a query that calculated the # of pages on each segment, and summed them up. Now, it runs a query that sums up the table sizes in bytes from each segment, and divides by BLCKSZ to get relpages. For a heap table, there's no difference, because the size in bytes is always divisible by BLCKSZ, but for AO tables, you can get a slightly different result. 

```
postgres=# select gp_segment_id, pg_relation_size('incr_analyze_test_1_prt_2') from gp_dist_random('gp_id');
 gp_segment_id | pg_relation_size 
---------------+------------------
             0 |            11064
             1 |            12024
             2 |            11576
(3 rows)
```

hlinnaka commented:
I fixed the last regression failures, and pushed. This has changed somewhat since the initial version in the PR, but the design and code structure is the same, so I think this has been reviewed enough. But if you've got time, feel free to review the final version that was pushed one more time. Thanks for the reviews!
